src/algebra/group/prod.lean

```diff
@@ -0,0 +1,236 @@
+@[to_additive]
+instance [has_inv M] [has_inv N] : has_inv (M × N) := ⟨λp, (p.1⁻¹, p.2⁻¹)⟩
+
+@[simp, to_additive]
+lemma fst_inv [has_inv G] [has_inv H] (p : G × H) : (p⁻¹).1 = (p.1)⁻¹ := rfl
+@[simp, to_additive]
+lemma snd_inv [has_inv G] [has_inv H] (p : G × H) : (p⁻¹).2 = (p.2)⁻¹ := rfl
+@[to_additive]
+lemma inv_mk [has_inv G] [has_inv H] (a : G) (b : H) : (a, b)⁻¹ = (a⁻¹, b⁻¹) := rfl
+
+instance [add_semigroup A] [add_semigroup B] : add_semigroup (A × B) :=
+{ add_assoc := assume a b c, mk.inj_iff.mpr ⟨add_assoc _ _ _, add_assoc _ _ _⟩,
+  .. prod.has_add }
+@[to_additive add_semigroup]
+instance [semigroup M] [semigroup N] : semigroup (M × N) :=
+{ mul_assoc := assume a b c, mk.inj_iff.mpr ⟨mul_assoc _ _ _, mul_assoc _ _ _⟩,
+  .. prod.has_mul }
+
+instance [add_monoid A] [add_monoid B] : add_monoid (A × B) :=
+{ zero_add := assume a, prod.rec_on a $ λa b, mk.inj_iff.mpr ⟨zero_add _, zero_add _⟩,
+  add_zero := assume a, prod.rec_on a $ λa b, mk.inj_iff.mpr ⟨add_zero _, add_zero _⟩,
+  .. prod.add_semigroup, .. prod.has_zero }
+@[to_additive add_monoid]
+instance [monoid M] [monoid N] : monoid (M × N) :=
+{ one_mul := assume a, prod.rec_on a $ λa b, mk.inj_iff.mpr ⟨one_mul _, one_mul _⟩,
+  mul_one := assume a, prod.rec_on a $ λa b, mk.inj_iff.mpr ⟨mul_one _, mul_one _⟩,
+  .. prod.semigroup, .. prod.has_one }
+
+instance [add_group A] [add_group B] : add_group (A × B) :=
+{ add_left_neg := assume a, mk.inj_iff.mpr ⟨add_left_neg _, add_left_neg _⟩,
+  .. prod.add_monoid, .. prod.has_neg }
+@[to_additive add_group]
+instance [group G] [group H] : group (G × H) :=
+{ mul_left_inv := assume a, mk.inj_iff.mpr ⟨mul_left_inv _, mul_left_inv _⟩,
+  .. prod.monoid, .. prod.has_inv }
+
+instance [add_comm_semigroup A] [add_comm_semigroup B] : add_comm_semigroup (A × B) :=
+{ add_comm := assume a b, mk.inj_iff.mpr ⟨add_comm _ _, add_comm _ _⟩,
+  .. prod.add_semigroup }
+@[to_additive add_comm_semigroup]
+instance [comm_semigroup G] [comm_semigroup H] : comm_semigroup (G × H) :=
+{ mul_comm := assume a b, mk.inj_iff.mpr ⟨mul_comm _ _, mul_comm _ _⟩,
+  .. prod.semigroup }
+
+instance [add_comm_monoid A] [add_comm_monoid B] : add_comm_monoid (A × B) :=
+{ .. prod.add_comm_semigroup, .. prod.add_monoid }
+@[to_additive add_comm_monoid]
+instance [comm_monoid M] [comm_monoid N] : comm_monoid (M × N) :=
+{ .. prod.comm_semigroup, .. prod.monoid }
+
+instance [add_comm_group A] [add_comm_group B] : add_comm_group (A × B) :=
+{ .. prod.add_comm_semigroup, .. prod.add_group }
+@[to_additive add_comm_group]
+instance [comm_group G] [comm_group H] : comm_group (G × H) :=
+{ .. prod.comm_semigroup, .. prod.group }
+
+end prod
+
+namespace monoid_hom
+
+variables (M N) [monoid M] [monoid N]
+
+/-- Given monoids `M`, `N`, the natural projection homomorphism from `M × N` to `M`.-/
+@[to_additive "Given additive monoids `A`, `B`, the natural projection homomorphism
+from `A × B` to `A`"]
+def fst : M × N →* M := ⟨prod.fst, rfl, λ _ _, rfl⟩
```

Comment:
`monoid_hom.fst` etc were moved from `prod.` namespace to allow dot syntax.

Reply:
As far as I can see we're not going to use anything like `f.fst`. But nevertheless this is a more sensible namespace to find this construction in.